### ISSUE TITLE
Training from scratch for English model, somewhat listenable after 180k updates.

### ISSUE THREAD

**Issue by Quentin1168** [SWivid/F5-TTS#548]
### Checks

- [x] This template is only for usage issues encountered.
- [X] I have thoroughly reviewed the project documentation but couldn't find information to solve my problem.
- [X] I have searched for existing issues, including closed ones, and couldn't find a solution.
- [X] I confirm that I am using English to submit this report in order to facilitate communication.

### Environment Details

pytorch:2.4.0
python:3.11
cuda:12.4
ubuntu:22.04

### Steps to Reproduce

I am trying to train a English model from scratch. I have ~400 hours of audio data from the Libri TTS R dataset. Combining the dev_clean, test_clean and train_clean_360 data together. 

I am currently renting a Single A100 SXM 80 GB vram GPU. 

Regarding Training Params:
Initial LR: 0.000075
Batch Size is 38400 on 1 GPU
Batch Size Type is Frame.
Max Samples is 64.
Gradient Acumulation Steps is 1 
Max Gradient Norm is 1
Warmup Updates is 10k
Using Bf16 precision.

### ✔️ Expected Behavior

Having seen some other benchmarks using way smaller datasets and different batch sizes, I expected my results to be better than what I currently have.

The audio is jumbled and gibberish in most parts with some intelligble phrases being heard clearly, mostly at the end.

An example:
Text to be generated.
"How is this still doing this badly? I don't know what is going on. I told my boss I'll get this to him by Monday but I don't think it's going to happen."
Result:
https://voca.ro/19IKC1qVps3k

This is the loss and lr graphs.
![Screenshot 2024-11-29 024328](https://github.com/user-attachments/assets/54ffd6cf-ebd8-4d7e-abaa-faab99df1572)
![image](https://github.com/user-attachments/assets/37f68e6a-b1b3-4914-a7db-f07afd4a7990)

Would I need to train longer? I am currently at 190k steps and still training. The loss has stopped going down for a few hours, would that mean my model will cease improving?
I am new to this, and only was able to compile this training setup through seeing the guides here. I would greatly appreciate any help regarding to this issue.

**Comment by SWivid** [comment]
the loss is fine, it's normal after warmup and several k steps, it gets into a phase that just decreases at a very slow rate

we haven't try libritts-r, but tried libritts

we train with small model (~155M), 38400 * 8GPU
200K to have pretty good results, 500K is OK to stop with impressive performance on English test set (with a certain zero-shot ability but obviously inferior to our base mode trained with more diverse data).

so, just train longer, will be better

**Comment by Quentin1168** [comment]
> the loss is fine, it's normal after warmup and several k steps, it gets into a phase that just decreases at a very slow rate
> 
> we haven't tried libritts-r, but tried libritts
> 
> trained with small model (~155M), 38400 * 8GPU 200K to have pretty good results, 500K is OK to stop with impressive performance on English test set (with a certain zero-shot ability but obviously inferior to our base mode trained with more diverse data).
> 
> so, just train longer, will be better

Hey, thank you so much for your reply. How would I change the size of my model? I have not touched anything regarding the model size so I assume I am training on the large one. Thank you.

**Comment by SWivid** [comment]
If with the latest version check the readme for training, simply leverage Hydra config yamls

Otherwise in the older version, is to change the hyperparams in train.py, we have mentioned small model config in our paper, and you can also find in under scripts/ the file including model params

**Comment by ZhikangNiu** [comment]
@Quentin1168 Based on my experimental results, I do not recommend using the LibriTTS-R dataset; instead, I suggest using the LibriTTS dataset.

**Comment by Quentin1168** [comment]
> @Quentin1168 Based on my experimental results, I do not recommend using the LibriTTS-R dataset; instead, I suggest using the LibriTTS dataset.

I thought the R one would have clearer speech. May I ask why? Just wanted to understand.

**Comment by Quentin1168** [comment]
> I'm by no means an expert and it may be best to hear from the project authors, whom I'm sure will also chime in. This model from my reading requires quite a bit of training initially, particularly if you're not finetuning. As someone who's also interested in training from scratch, I'm interested how you get on with this. Issue #509 is where I got a lot of the info on the amount of training required. Hope this helps a bit.

Hey, since you are interested in the progress, here is how it is going at 280k updates. 

https://voca.ro/1iQKdblP7fM4

It's a bit better from the one at 200k updates, but still a bit jumbly. I think it does better at very short sentences.

**Comment by ZhikangNiu** [comment]
> > @Quentin1168 Based on my experimental results, I do not recommend using the LibriTTS-R dataset; instead, I suggest using the LibriTTS dataset.
> 
> I thought the R one would have clearer speech. May I ask why? Just wanted to understand.

Enhanced audio may sound better to the human ear, but it may be detrimental to the model. The results I obtained using LibriTTS-R were quite poor in WER and SIM

**Comment by leoiania** [comment]
Thank you @Quentin1168 for sharing your results. Can I ask you how long did it take for you to get to 280k updates on a A100 SXM 80 GB vram ? It may be a useful information

**Comment by ZhikangNiu** [comment]
H100*8 Libritts all train datasets and F5 small update 300k steps will need 24hours，so the time will double if a100*8发自我的 iPhone在 2024年11月30日，07:07，Leonardo Iania ***@***.***> 写道：﻿
Thank you @Quentin1168 for sharing your results. Can I ask you how long did it take for you to get to 280k updates on a A100 SXM 80 GB vram ? It may be a useful information

**Comment by Quentin1168** [comment]
Alright, here's an update. I switched to the small model and have been training with the same parameters and cfg_strength = 3.0 (modified in utils_infer)

Here is the loss graph:

![image](https://github.com/user-attachments/assets/e25a9523-5b7c-421a-b0a3-224a31252ad5)

Here are some samples of the same sentence above:

https://voca.ro/1mBNluWVXkki
https://voca.ro/1iahPbxu1vcr

I'll probably follow Zhikang's suggestion of using the normal LibriTTS but I have limited resources with using this hardware. Should I wait it out a bit more or just stop and retrain with normal LibriTTS later?

**Comment by rlenain** [comment]
@Quentin1168 is your suggestion to use a smaller model because of the smaller size of the dataset?
Do you have a rough idea of the amount of data required to train with the larger model size, apart from what is being said in the paper?

**Comment by ILG2021** [comment]
Hello, everyone, I am trying to train a 44k model from scratch too, I have about 10 hours data, 2000 batch on a single 4080 GPU. after 250k updates, the voice become clear, but content is a mess. When go to 300k, voice become some noisy. The learning rate is 1e-13, gradient is nearly not update. I think the model can not learn the alignment.
When I use pretrain 22k model, 14k will get a clear voice and accurate content at the same parameters.

**Comment by ndhuynh02** [comment]
I am training also training F5-base on LibriTTS, with batch size of 19200 and 16 grad accumulation on a single 80GB A100 GPU. However, after 1M steps, the content is still not correct like in https://github.com/SWivid/F5-TTS/issues/509.
BTW, I am using the exact provided code for LibriTTS pre-processing.

**Comment by ZhikangNiu** [comment]
> BTW, I am using the exact provided code for LibriTTS pre-processing.

My setting is batch size=38400 and 1 grad accumulation with 4GPU, and can get not bad result after 300k

**Comment by ILG2021** [comment]
Anyone tried 4x4090? A100 and H100 is too expensive for me.

**Comment by ILG2021** [comment]
> My setting is batch size=38400 and 1 grad accumulation with 4GPU, and can get not bad result after 300k.
> Model setting is F5TTS Small

Which GPU do you use? and how long do you cost? How many epochs do you set?

**Comment by Ming-er** [comment]
> > BTW, I am using the exact provided code for LibriTTS pre-processing.
> 
> My setting is batch size=38400 and 1 grad accumulation with 4GPU, and can get not bad result after 300k. Model setting is F5TTS Small

Would you mind provide a detailed evaluation result on Librispeech using this small model trained on LibriTTS?

**Comment by ndhuynh02** [comment]
I do wonder, @ZhikangNiu , are you actually training the model from scratch for are you finetuning the model? Because I have trained F5Small on LibriTTS for 1M steps, and it is not that great of a result

**Comment by ZhikangNiu** [comment]
1. My training config is listed as follows, and I trained all of F5TTS_Small on 4*H100 with fp16.
2. All of them are trained from scratch @ndhuynh02 
3. The data preprocess script is https://github.com/SWivid/F5-TTS/blob/main/src/f5_tts/train/datasets/prepare_libritts.py
4. I always evaluate them on LibriSpeech-PC test-clean and I think seedtts-eval is difficult for small model, nfe step=32. The results on 300k look good. @ILG2021 
5. We don't try 8-bit Adam Optimizer @ILG2021 
```yaml
hydra:
  run:
    dir: ckpts/${model.name}_${model.mel_spec.mel_spec_type}_${model.tokenizer}_${datasets.name}
ckpts:
  logger: wandb  # wandb | tensorboard | None
  exp_name: ${model.name}_${model.mel_spec.mel_spec_type}_${model.tokenizer}_${datasets.name}
  save_per_updates: 50000  # save checkpoint per steps
  last_per_steps: 5000  # save last checkpoint per steps
  save_dir: ckpts/${model.name}_${model.mel_spec.mel_spec_type}_${model.tokenizer}_${datasets.name}

datasets:
  name: LibriTTS_100_360_500
  batch_size_per_gpu: 38400  # 8 GPUs, 8 * 38400 = 307200
  batch_size_type: frame  # "frame" or "sample"
  max_samples: 64  # max sequences per batch if use frame-wise batch_size. we set 32 for small models, 64 for base models
  num_workers: 16

optim:
  epochs: LibriTTS_100_360_500
  learning_rate: 7.5e-5
  num_warmup_updates: 20000  # warmup steps
  grad_accumulation_steps: 1  # note: updates = steps / grad_accumulation_steps
  max_grad_norm: 1.0  # gradient clipping
  bnb_optimizer: False  # use bnb 8bit AdamW optimizer or not

model:
  name: F5TTS_Small
  tokenizer: pinyin
  tokenizer_path: None  # if tokenizer = 'custom', define the path to the tokenizer you want to use (should be vocab.txt)
  frac_lengths_mask: [0.7,1.0]
  arch:
    dim: 768
    depth: 18
    heads: 12
    ff_mult: 2
    text_dim: 512
    conv_layers: 4
    refine_type: "conv"
    checkpoint_activations: False  # recompute activations and save memory for extra compute
  mel_spec:
    target_sample_rate: 24000
    n_mel_channels: 100
    hop_length: 256
    win_length: 1024
    n_fft: 1024
    mel_spec_type: vocos  # 'vocos' or 'bigvgan'
  vocoder:
    is_local: False  # use local offline ckpt or not
    local_path: None  # local vocoder path
```
accelerate config
```yaml
compute_environment: LOCAL_MACHINE
debug: false
distributed_type: MULTI_GPU
downcast_bf16: 'no'
enable_cpu_affinity: true
gpu_ids: all
machine_rank: 0
main_training_function: main
mixed_precision: fp16
num_machines: 1
num_processes: 4
rdzv_backend: static
same_network: true
tpu_env: []
tpu_use_cluster: false
tpu_use_sudo: false
use_cpu: false
```

**Comment by ZhikangNiu** [comment]
I re-open this issue; feel free to discuss

**Comment by ZhikangNiu** [comment]
> Anyone tried 4x4090? A100 and H100 is too expensive for me. Could I open 8 bit adam to speed up?

If you try 4x4090 and want to keep same size as our, you can set checkpoint_activations=True

**Comment by ILG2021** [comment]
I have trained 6400k on a single 4080, 4000 batch size, I use f5 base config. The result is still not good, sound with noise(I have train another model with bigvgan 24k, still has some noise, I don't know if it related by the vocoder) and not align well.  so what's the possible problem, it is because the batch size too small or train step is still not enough?
38400*4*300 /4000  = 11520 k?
I see someone use 8 h100 train for 100k, 
38400*8*100/4000 = 7680k nearly 6400k. So I think the step is enough. 
So is it cause by the vocoder or batch size?

@ZhikangNiu

**Comment by ILG2021** [comment]
> [@Ming-er](https://github.com/Ming-er) [@ILG2021](https://github.com/ILG2021) [@ndhuynh02](https://github.com/ndhuynh02) [@rlenain](https://github.com/rlenain) [@Quentin1168](https://github.com/Quentin1168) [@leoiania](https://github.com/leoiania) I strongly recommend that you use the code from version 1.0.0. We have employed zero init adaln, full rope and text mask padding to make it converge faster during training and achieve better results, even on smaller models and smaller dataset.

Recently I am trying bigvgan 44k with v1 base, but the result is seems not as expect. I use 4000 batch size accumulation step 9, so the equal batch size = 36000, after 180k updates, the generate audio is still contains many noise. Another community developer has success train from scratch with one A100, batch size 38400 accumulation 1, vocos 24k, v1 base, in his video, his model is clear after 100k updates ( ). So what's may be the problem? The ref audio 44k is ok, so I don't think the vocoder related code has problem.

@ZhikangNiu @SWivid
